jena-core/src/main/java/org/apache/jena/util/PrefixMappingUtils.java

```diff
@@ -0,0 +1,362 @@
+        Consumer<Node> process = (node) -> {
+            if ( ! node.isURI() )
+                return ;
+            String uri = node.getURI() ;
+            
+            int idx = SplitIRI.splitpoint(uri) ;
+            if ( idx < 0 )
+                return ;
+            String nsURI = SplitIRI.namespaceTTL(uri) ;
+            String prefix = prefixMapping.getNsURIPrefix(nsURI) ;
+            if ( prefix != null )
+                inUse.add(prefix) ;
+        } ;
+        
+        Iterator<Triple> iter = graph.find(null, null, null) ;
+        while(iter.hasNext()) {
+            Triple triple = iter.next() ;
+            process.accept(triple.getSubject()) ; 
+            process.accept(triple.getPredicate()) ;
+            process.accept(triple.getObject()) ; 
+            if ( inUse.size() == prefixURIs.size() )
+                // Fast exit.
+                break ;
+        }
+        
+        if ( pmap.size() == inUse.size() )
+            return prefixMapping ;
+
+        // Build result.
+        PrefixMapping pmap2 = new PrefixMappingImpl() ;
+        inUse.forEach((prefix)-> pmap2.setNsPrefix(prefix, prefixMapping.getNsPrefixURI(prefix)) ) ;
+        return pmap2 ;
+    }
+    
+    /** Check every URI as a possible use of a prefix */ 
+    private static Set<String> fullMethod(Model m) {
+        /* Method: Covers prefixes not based on "/", "#" or final ":" splitting.  
+         * 
+         * Build a trie to use as a partial lookup matcher.
+         * For each URI in the data, look it up as a partial match in the trie
+         * to get all uris in the prefix map that apply.  
+         */
+        
+        // Map prefix to URI.
+        Map<String, String> pmap = m.getNsPrefixMap() ;
+        // Map URI to prefix, with partial lookup (all uri keys that partly match the URI) 
+        Trie<String> trie = new Trie<>() ;                      
+        
+        // change to add(uri, prefix) to get prefixes.
+        pmap.forEach((prefix,uri)-> trie.add(uri, uri)) ;   
+        
+        Iterator<Triple> iter = m.getGraph().find(null, null, null) ;
+        // Prefix URIs in use.
+        Set<String> inUseURIs = new HashSet<>() ;
+        while(iter.hasNext()) {
+            Triple triple = iter.next() ;
+            processFull(trie, inUseURIs, triple.getSubject()) ; 
+            processFull(trie, inUseURIs, triple.getPredicate()) ;
+            processFull(trie, inUseURIs, triple.getObject()) ; 
+            if ( pmap.size() == inUseURIs.size() )
+                break ;
+        }
+        return inUseURIs ;
+    }
+
+    private static void processFull(Trie<String> trie, Set<String> prefixesInUse, Node node) {
+        if ( ! node.isURI() )
+            return ;
+        String uri = node.getURI() ;
+        // Shorten to "/" or "#" or ":"
+        String pref = uri ;
+        
+        // Get all under the pref
+        List<String> hits = trie.partialSearch(pref) ;
+        if ( hits == null || hits.isEmpty() )
```

Comment:
Looking at `Trie::partialSearch`, I don't think `hits` can be `null` here. I guess why you don't check for `null` above?

Reply:
Just being defensive.